src/data/parquet_parser.h

```diff
@@ -0,0 +1,189 @@
+/*!
+ *  Copyright (c) 2021 by Contributors
+ * \file parquet_parser.h
+ * \brief iterator parser to parse parquet format
+ * \author Chengyang Gu
+ */
+#ifndef DMLC_DATA_PARQUET_PARSER_H_
+#define DMLC_DATA_PARQUET_PARSER_H_
+
+#include <dmlc/data.h>
+#include <dmlc/strtonum.h>
+#include <dmlc/parameter.h>
+#include <cmath>
+#include <cstring>
+#include <map>
+#include <string>
+#include <limits>
+#include <future>
+#include <algorithm>
+#include <memory>
+#include <vector>
+#include "../data/row_block.h"
+#include "../data/parser.h"
+#include "arrow/io/api.h"
+#include "parquet/api/reader.h"
+
+
+namespace dmlc {
+namespace data {
+
+struct ParquetParserParam : public Parameter<ParquetParserParam> {
+  std::string format;
+  int label_column;
+  int weight_column;
+  int nthreads;
+
+  DMLC_DECLARE_PARAMETER(ParquetParserParam) {
+    DMLC_DECLARE_FIELD(format).set_default("parquet")
+      .describe("File format.");
+    DMLC_DECLARE_FIELD(label_column).set_default(0)
+      .describe("Column index (0-based) that will put into label.");
+    DMLC_DECLARE_FIELD(weight_column).set_default(-1)
+      .describe("Column index that will put into instance weights.");
+    DMLC_DECLARE_FIELD(nthreads).set_default(2)
+      .describe("Column index that will put into instance weights.");
+  }
+};
+
+template <typename IndexType, typename DType = real_t>
+class ParquetParser : public ParserImpl<IndexType, DType> {
+ public:
+  ParquetParser(const std::string& filename,
+                const std::map<std::string, std::string>& args,
+                int nthread) : row_groups_read_(0), nthread_(nthread) {
```

Comment:
Is the `nthread` necessary?  It's passed as `-1`, then `nthread_` is overridden by the one in `param_`.

Reply:
I agree, It's not. In fact, `nthread` first being passed as `-1` is unnecessary. Furthermore, I observe that nthreads for other parsers has been hardcoded as 2 here https://github.com/dmlc/dmlc-core/blob/main/src/data.cc#L59 (also L29 and L44). However, as @hcho3 has pointed out, the effective nthread is defined as `omp_get_max_threads()` here https://github.com/dmlc/dmlc-core/blob/main/src/data/text_parser.h#L114. Maybe we can remove `nthread` for all the parsers. 

In my practice, CPU usage never exceeded 500% on a 64-core CPU, which explains why this bug went unnoticed.

Reply:
> nthread first being passed as -1 is unnecessary

Thanks for the confirmation.

>  Maybe we can remove nthread for all the parsers and allow users to specify it in the URI.

For now we can limit the scope of this PR and just get the parser in first.  The nthread is quite problematic from my experience.  Using more threads means spawning more batches and buffers, leading to memory usage issues and reproducibility issues.  Can we remove this `-1`? and `nthread` in constructor?  Sorry for the inconvience.

Reply:
Thanks for your feedback! I've removed the `nthread` argument and changed the default `nthreads` (passed by URI) to 1 [here](https://github.com/bridgream/dmlc-core/blob/parquet_support/src/data/parquet_parser.h#L44). I use `nthreads` in my project and it has been good so far (with predictable memory usage and reproducible results). Hopefully we can expose this feature to users in the future.